enterprise/internal/batches/store/worker_workspace_execution.go

```
@@ -129,12 +131,18 @@ func (s *batchSpecWorkspaceExecutionWorkerStore) markFinal(ctx context.Context,
 		return false, err
 	}
 	defer func() {
+		// If no matching record was found, revert the tx.
+		if !ok && err == nil {
+			tx.Done(errors.New("record not found"))
```

```suggestion
			tx.Done(errors.Newf("record %d not found", id))
```

---

the err is lost anyways, we just don't expose tx.Rollback() directly 😬